### PR TITLE
Detect then/catch calls that should never be called

### PR DESCRIPTION
**Pull request by weikinhuang**

### CODE REVIEW

File: src/promises.js [hunk 1]
```diff
@@ -17,42 +18,67 @@ const WATCHED_PROMISE_IMPLEMENTATIONS = new Map();
...
+      method
+    };
+    strayPromises.push(promiseData);
+
+    const newPromise = thenablePrototype.apply(
       this,
       args.map((fn) => rebindResolver(fn, localIdx))
     );
+
+    // keep an handy call stack reference
```
Comment (kaicataldo) [inline src/promises.js]: Typo: `an` -> `a`

File: src/promises.js [hunk 2]
```diff
@@ -146,15 +185,52 @@ export function setupPromiseDetection() {
...
+  isCleaningUp = true;
 
   // reset timer cache for next test
   strayPromises = [];
 
   if (!this.__strayPromisesIgnored && localStrayPromises.length > 0) {
-    let firstStrayPromise = localStrayPromises.shift();
-    throw firstStrayPromise.err;
+    let unresolvedPromises = [...localStrayPromises].filter(({ hasBeenCalled }) => !hasBeenCalled);
+    // debugger;
```
Comment (kaicataldo) [inline src/promises.js]: Forgotten comment?

**Comment by kaicataldo** [comment]
LGTM

**Review by kaicataldo — APPROVED** [review]

File: src/promises.js [hunk 3]
```diff
@@ -17,42 +18,67 @@ const WATCHED_PROMISE_IMPLEMENTATIONS = new Map();
...
+        return false;
+      }
+      strayPromise.hasBeenCalled = true;
+      return true;
+    });
     return fn.apply(this, arguments);
-  };
+  }
+  // this is useful for debugging
+  // reboundResolver.originalFn = String(fn);
```
Comment (nemtsov) [inline src/promises.js:35]: (Not a blocker) Considering enabling this on `/jasmine-stray-promises/.test(process.env.NODE_DEBUG)` instead (or something similar).

**Review by nemtsov — APPROVED** [review]

File: src/promises.js (hunk 3 above)
Comment (weikinhuang) [reply src/promises.js:35]: that's a good call, I'll do that in the documentation PR next

**Pull request merged**